libgnucash/app-utils/gnc-quotes.cpp

```diff
@@ -0,0 +1,1056 @@
+    const QuoteSources& sources() noexcept { return m_sources; }
+    GList* sources_as_glist ();
+    bool had_failures() noexcept { return !m_failures.empty(); }
+    const QFVec& failures() noexcept;
+    std::string report_failures() noexcept;
+
+private:
+    std::string query_fq (const char* source, const StrVec& commoditites);
+    std::string query_fq (const CommVec&);
+    bpt::ptree parse_quotes (const std::string& quote_str);
+    void create_quotes(const bpt::ptree& pt, const CommVec& comm_vec);
+    std::string comm_vec_to_json_string(const CommVec&) const;
+    GNCPrice* parse_one_quote(const bpt::ptree&, gnc_commodity*);
+
+    std::unique_ptr<GncQuoteSource> m_quotesource;
+    std::string m_version;
+    QuoteSources m_sources;
+    QFVec m_failures;
+    QofBook *m_book;
+    gnc_commodity *m_dflt_curr;
+};
+
+class GncFQQuoteSource final : public GncQuoteSource
+{
+    const bfs::path c_cmd;
+    const std::string c_fq_wrapper;
+    bool m_ready;
+    std::string m_version;
+    StrVec m_sources;
+    std::string m_api_key;
+public:
+    GncFQQuoteSource();
+    ~GncFQQuoteSource() = default;
+    virtual const std::string& get_version() const noexcept override { return m_version; }
+    virtual const StrVec& get_sources() const noexcept override { return m_sources; }
+    virtual QuoteResult get_quotes(const std::string&) const override;
+    virtual bool usable() const noexcept override { return m_ready; }
+private:
+    QuoteResult run_cmd (const StrVec& args, const std::string& json_string) const;
+
+};
+
+static void show_quotes(const bpt::ptree& pt, const StrVec& commodities, bool verbose);
+static void show_currency_quotes(const bpt::ptree& pt, const StrVec& commodities, bool verbose);
+
+static const std::string empty_string{};
+
+GncFQQuoteSource::GncFQQuoteSource() :
+c_cmd{bp::search_path("perl")},
+c_fq_wrapper{std::string(gnc_path_get_bindir()) + "/finance-quote-wrapper"},
+m_ready{false},
+m_version{}, m_sources{}, m_api_key{}
+{
+    StrVec args{"-w", c_fq_wrapper, "-v"};
+    const std::string empty_string;
+    auto [rv, sources, errors] = run_cmd(args, empty_string);
+    if (rv)
+    {
+        std::string err{bl::translate("Failed to initialize Finance::Quote: ")};
+        for (auto err_line : errors)
+            err += err_line.empty() ? "" : err_line + "\n";
+        throw(GncQuoteSourceError(err));
+    }
+    if (!errors.empty())
+    {
+        std::string err{bl::translate("Finance::Quote check returned error ")};
+        for(const auto& err_line : errors)
+            err += err.empty() ? "" : err_line + "\n";
+        throw(GncQuoteSourceError(err));
+    }
+    static const boost::regex version_fmt{"[0-9]\\.[0-9][0-9]"};
+    auto version{sources.front()};
+    if (version.empty() || !boost::regex_match(version, version_fmt))
+    {
+        std::string err{bl::translate("Invalid Finance::Quote Version ")};
+            err +=  version.empty() ? "" : version;
```

Comment:
Interesting. Did you actually run into the situation where F::Q returned an invalid version number ?

Reply:
Nope, I just copied this from price-quotes.scm. But we don't even care what version string F::Q returns as long as it returns something, so I've deleted the regex and the check.